dask/array/creation.py

```diff
@@ -501,6 +501,81 @@ def diag(v):
     return Array(graph, name, (chunks_1d, chunks_1d), dtype=v.dtype)
 
 
+@wraps(np.diagonal)
+def diagonal(a, offset=0, axis1=0, axis2=1):
+    name = 'diagonal-' + tokenize(a)
+
+    if a.ndim < 2:
+        # NumPy uses `diag` as we do here.
+        raise ValueError("diag requires an array of at least two dimensions")
+
+    if axis1 == axis2:
+        raise ValueError("axis1 and axis2 cannot be the same")
+
+    if offset >= 0:
+        dim1_offset = 0
+        dim2_offset = +offset
+    else:
+        dim1_offset = -offset
+        dim2_offset = 0
+
+    if axis1 > axis2:
+        axis1, axis2 = axis2, axis1
+        dim1_offset, dim2_offset = dim2_offset, dim1_offset
+        offset *= -1
+
+    left_idx = list(range(len(a.shape)))
+    del left_idx[axis2]
+    del left_idx[axis1]
+
+    left_shape = tuple([a.shape[i] for i in left_idx])
+    right_shape = (max(min(a.shape[axis1] - dim1_offset,
+                           a.shape[axis2] - dim2_offset), 0),)
+    shape = left_shape + right_shape
+
+    if isinstance(a, np.ndarray):
```

Comment:
I think you can skip this block by just casting numpy arrays to dask arrays (i.e., with `dask.array.asarray()`) and proceed with the normal dask logic.